pkg/controllers/dnspolicy/dnspolicy_dnsrecords.go

```diff
@@ -137,3 +135,81 @@ func (r *DNSPolicyReconciler) deleteGatewayDNSRecords(ctx context.Context, gatew
 	}
 	return nil
 }
+
+func (r *DNSPolicyReconciler) buildClusterGateway(ctx context.Context, downstreamClusterName string, clusterAddress []gatewayv1beta1.GatewayAddress) (dns.ClusterGateway, error) {
+	var target dns.ClusterGateway
+	singleClusterAddresses := make([]gatewayv1beta1.GatewayAddress, len(clusterAddress))
+
+	mc := &clusterv1.ManagedCluster{}
+	if err := r.Client().Get(ctx, client.ObjectKey{Name: downstreamClusterName}, mc, &client.GetOptions{}); err != nil {
+		return target, err
+	}
+
+	for i, addr := range clusterAddress {
+		addrType := gatewayv1beta1.IPAddressType
+		if *addr.Type == gateway.MultiClusterHostnameAddressType {
+			addrType = gatewayv1beta1.HostnameAddressType
+		}
+
+		singleClusterAddresses[i] = gatewayv1beta1.GatewayAddress{
+			Type:  &addrType,
+			Value: addr.Value,
+		}
+	}
+
+	target = *dns.NewClusterGateway(mc, singleClusterAddresses)
+
+	return target, nil
+}
+
+func getClusterGatewayAddresses(gw *gatewayv1beta1.Gateway) map[string][]gatewayv1beta1.GatewayAddress {
+	clusterAddrs := make(map[string][]gatewayv1beta1.GatewayAddress, len(gw.Status.Addresses))
+
+	for _, address := range gw.Status.Addresses {
+		var gatewayAddresses []gatewayv1beta1.GatewayAddress
+
+		//Default to Single Cluster (Normal Gateway Status)
+		cluster := "none"
+		addressValue := address.Value
+
+		//Check for Multi Cluster (MGC Gateway Status)
+		if *address.Type == gateway.MultiClusterIPAddressType || *address.Type == gateway.MultiClusterHostnameAddressType {
+			tmpCluster, tmpAddress, found := strings.Cut(address.Value, "/")
+			//If this fails something is wrong and the value hasn't been set correctly
+			if found {
+				cluster = tmpCluster
+				addressValue = tmpAddress
+			}
+		}
+
+		gatewayAddresses = append(gatewayAddresses, gatewayv1beta1.GatewayAddress{
+			Type:  address.Type,
+			Value: addressValue,
+		})
+		clusterAddrs[cluster] = gatewayAddresses
+	}
+
+	return clusterAddrs
+}
+
+func listenerTotalAttachedRoutes(upstreamGateway *gatewayv1beta1.Gateway, downstreamCluster, specListenerName string, addresses []gatewayv1beta1.GatewayAddress) int {
+	for _, listener := range upstreamGateway.Status.Listeners {
+		// assuming all adresses of the same type on the gateway
+		// for Multi Cluster (MGC Gateway Status)
+		if *addresses[0].Type == gateway.MultiClusterIPAddressType || *addresses[0].Type == gateway.MultiClusterHostnameAddressType {
+			clusterName, listenerName, found := strings.Cut(string(listener.Name), ".")
+			if !found {
+				return 0
+			}
+			if clusterName == downstreamCluster && (listenerName == specListenerName || specListenerName == "wildcard") {
```

Comment:
What is the check for "wildcard" for?

Reply:
To allow for the wildcard listeners. I could also check for `*` in the hostname :thinking:  And I think I'll do that. And could finally justify the use of the `contains` for checking names and `*` symbols in the listener